File: course_1/module_2/examples/closest_pair.py
from math import sqrt
from typing import List, Tuple
import bisect


class ClosestPair:
    def __init__(self, points: List[Tuple[float, float]]):
        """
        Initializes the object with a given list of points.
        """
        self.points = points

    @staticmethod
    def euclidean_distance(p1: Tuple[float, float], p2: Tuple[float, float]) -> float:
        """
        Calculates the Euclidean distance between two points.
        """
        return sqrt((p1[0] - p2[0]) ** 2 + (p1[1] - p2[1]) ** 2)

    def brute_force(self, points: List[Tuple[float, float]]) -> Tuple[
        float, Tuple[Tuple[float, float], Tuple[float, float]]]:
        """
        Solves the closest pair problem using brute-force for a small set of points.
        """
        min_dist = float('inf')
        closest_pair = None
        n = len(points)

        for i in range(n):
            for j in range(i + 1, n):
                dist = self.euclidean_distance(points[i], points[j])
                if dist < min_dist:
                    min_dist = dist
                    closest_pair = (points[i], points[j])

        return min_dist, closest_pair

    def closest_split_pair(self, px: List[Tuple[float, float]], py: List[Tuple[float, float]], delta: float,
                           best_pair: Tuple[Tuple[float, float], Tuple[float, float]]) -> Tuple[
        float, Tuple[Tuple[float, float], Tuple[float, float]]]:
        """
        Handles the case where the closest pair of points is split between two halves.
        """
        mid_x = px[len(px) // 2][0]
        # Use binary search to select points within the strip.
        left = bisect.bisect_left(py, (mid_x - delta, float('-inf')))
        right = bisect.bisect_right(py, (mid_x + delta, float('inf')))
        sy = py[left:right]

        best_dist = delta
        len_sy = len(sy)

        # Check up to the next 7 points by y-coordinate.
        for i in range(len_sy):
            for j in range(1, min(8, len_sy - i)):
                p, q = sy[i], sy[i + j]
                dist = self.euclidean_distance(p, q)
                if dist < best_dist:
                    best_dist = dist
                    best_pair = (p, q)

        return best_dist, best_pair

    def closest_pair_recursive(self, px: List[Tuple[float, float]], py: List[Tuple[float, float]], left: int,
                               right: int) -> Tuple[float, Tuple[Tuple[float, float], Tuple[float, float]]]:
        """
        Recursive divide-and-conquer method to find the closest pair of points.
        """
        if right - left <= 3:
            # Use brute-force if the number of points is small.
            return self.brute_force(px[left:right])

        mid = (left + right) // 2
        midpoint_x = px[mid][0]

        # Divide py into Qy and Ry in one pass.
        Qy = []
        Ry = []
        for point in py:
            if point[0] <= midpoint_x:
                Qy.append(point)
            else:
                Ry.append(point)

        # Recursive calls for the left and right halves.
        dist_left, pair_left = self.closest_pair_recursive(px, Qy, left, mid)
        dist_right, pair_right = self.closest_pair_recursive(px, Ry, mid, right)

        # Choose the minimum distance from the two halves.
        if dist_left < dist_right:
            delta = dist_left
            best_pair = pair_left
        else:
            delta = dist_right
            best_pair = pair_right

        # Check the split pair case.
        dist_split, pair_split = self.closest_split_pair(px, py, delta, best_pair)

        # Return the smallest of the distances found.
        if delta < dist_split:
            return delta, best_pair
        else:
            return dist_split, pair_split

    def find_closest_pair(self) -> Tuple[float, Tuple[Tuple[float, float], Tuple[float, float]]]:
        """
        Finds the closest pair of points in the original set.
        """
        # Check if there are enough points.
        if len(self.points) < 2:
            raise ValueError("At least two points are required to find the closest pair.")

        # Sort points by x and y coordinates.
        px = sorted(self.points, key=lambda p: p[0])
        py = sorted(self.points, key=lambda p: p[1])

        # Start the recursive algorithm.
        return self.closest_pair_recursive(px, py, 0, len(px))


# Example usage and edge case testing.
if __name__ == "__main__":
    sample_points = [(2.1, 3.5), (1.1, 2.9), (3.6, 4.7), (4.2, 2.1), (0.9, 1.5)]
    closest_pair_solver = ClosestPair(sample_points)
    distance, pair = closest_pair_solver.find_closest_pair()
    print(f"The closest pair is {pair} with a distance of {distance:.2f}")

    # Edge case: fewer than two points.
    try:
        single_point = [(2.1, 3.5)]
        closest_pair_solver = ClosestPair(single_point)
        distance, pair = closest_pair_solver.find_closest_pair()
    except ValueError as e:
        print(e)

    # Edge case: large number of points (performance testing).
    import random

    large_points = [(random.uniform(-1000, 1000), random.uniform(-1000, 1000)) for _ in range(10000)]
    closest_pair_solver = ClosestPair(large_points)
    distance, pair = closest_pair_solver.find_closest_pair()
    print(f"Closest pair in large dataset: {pair} with a distance of {distance:.2f}")

File: course_1/module_2/examples/closest_pair.go
package main

import (
	"errors"
	"fmt"
	"math"
	"math/rand"
	"sort"
	"time"
)

type Point struct {
	X, Y float64
}

// ClosestPairSolver initializes the solver with the list of points.
type ClosestPairSolver struct {
	points []Point
}

func NewClosestPairSolver(points []Point) *ClosestPairSolver {
	return &ClosestPairSolver{points: points}
}

// EuclideanDistance calculates the Euclidean distance between two points.
func EuclideanDistance(p1, p2 Point) float64 {
	return math.Sqrt((p1.X-p2.X)*(p1.X-p2.X) + (p1.Y-p2.Y)*(p1.Y-p2.Y))
}

// BruteForce finds the closest pair using brute-force search.
func (solver *ClosestPairSolver) BruteForce(points []Point) (float64, [2]Point) {
	minDist := math.Inf(1)
	var closestPair [2]Point
	n := len(points)

	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			dist := EuclideanDistance(points[i], points[j])
			if dist < minDist {
				minDist = dist
				closestPair = [2]Point{points[i], points[j]}
			}
		}
	}
	return minDist, closestPair
}

// ClosestSplitPair checks for the closest pair across the divide.
func (solver *ClosestPairSolver) ClosestSplitPair(px, py []Point, delta float64, bestPair [2]Point) (float64, [2]Point) {
	midX := px[len(px)/2].X
	var sy []Point

	// Collect points within the delta-width vertical strip.
	for _, p := range py {
		if math.Abs(p.X-midX) < delta {
			sy = append(sy, p)
		}
	}

	bestDist := delta
	lenSy := len(sy)

	// Compare up to 7 closest neighbors.
	for i := 0; i < lenSy; i++ {
		for j := 1; j < 8 && i+j < lenSy; j++ {
			p, q := sy[i], sy[i+j]
			dist := EuclideanDistance(p, q)
			if dist < bestDist {
				bestDist = dist
				bestPair = [2]Point{p, q}
			}
		}
	}
	return bestDist, bestPair
}

// ClosestPairRecursive is a recursive divide-and-conquer method.
func (solver *ClosestPairSolver) ClosestPairRecursive(px, py []Point) (float64, [2]Point) {
	n := len(px)
	if n <= 3 {
		return solver.BruteForce(px)
	}

	mid := n / 2
	Qx := px[:mid]
	Rx := px[mid:]

	// Split py into Qy and Ry.
	var Qy, Ry []Point
	midpointX := px[mid].X
	for _, p := range py {
		if p.X <= midpointX {
			Qy = append(Qy, p)
		} else {
			Ry = append(Ry, p)
		}
	}

	// Recursive calls for left and right halves.
	distLeft, pairLeft := solver.ClosestPairRecursive(Qx, Qy)
	distRight, pairRight := solver.ClosestPairRecursive(Rx, Ry)

	// Choose the minimum distance from two halves.
	var delta float64
	var bestPair [2]Point
	if distLeft < distRight {
		delta = distLeft
		bestPair = pairLeft
	} else {
		delta = distRight
		bestPair = pairRight
	}

	// Check for closest split pair.
	distSplit, pairSplit := solver.ClosestSplitPair(px, py, delta, bestPair)

	// Return the smallest of the distances found.
	if delta < distSplit {
		return delta, bestPair
	} else {
		return distSplit, pairSplit
	}
}

// FindClosestPair initializes and runs the closest-pair search.
func (solver *ClosestPairSolver) FindClosestPair() (float64, [2]Point, error) {
	if len(solver.points) < 2 {
		return 0, [2]Point{}, errors.New("at least two points are required to find the closest pair")
	}

	// Sort points by X and Y coordinates.
	px := make([]Point, len(solver.points))
	py := make([]Point, len(solver.points))
	copy(px, solver.points)
	copy(py, solver.points)

	sort.Slice(px, func(i, j int) bool {
		return px[i].X < px[j].X
	})
	sort.Slice(py, func(i, j int) bool {
		return py[i].Y < py[j].Y
	})

	// Run recursive algorithm.
	dist, pair := solver.ClosestPairRecursive(px, py)
	return dist, pair, nil
}

func main() {
	rand.Seed(time.Now().UnixNano())
	points := []Point{{2.1, 3.5}, {1.1, 2.9}, {3.6, 4.7}, {4.2, 2.1}, {0.9, 1.5}}
	closestPairSolver := NewClosestPairSolver(points)
	distance, pair, err := closestPairSolver.FindClosestPair()
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	fmt.Printf("The closest pair is %v with a distance of %.2f\n", pair, distance)

	// Edge case: fewer than two points.
	singlePoint := []Point{{2.1, 3.5}}
	closestPairSolver = NewClosestPairSolver(singlePoint)
	distance, pair, err = closestPairSolver.FindClosestPair()
	if err != nil {
		fmt.Println("Error:", err)
	}

	// Performance testing with a large number of points.
	var largePoints []Point
	for i := 0; i < 10000; i++ {
		largePoints = append(largePoints, Point{
			X: (rand.Float64() * 2000) - 1000,
			Y: (rand.Float64() * 2000) - 1000,
		})
	}
	closestPairSolver = NewClosestPairSolver(largePoints)
	distance, pair, err = closestPairSolver.FindClosestPair()
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	fmt.Printf("Closest pair in large dataset: %v with a distance of %.2f\n", pair, distance)
}
